server/src/service/collection.rs

```diff
@@ -0,0 +1,201 @@
+
+use crate::service::item;
+use crate::KEYRING;
+
+#[derive(Default, Debug, Deserialize)]
+pub struct Collection {
+    pub items: Vec<item::Item>,
+    label: String,
+    locked: bool,
+    created: u64, // how to represent date
+    modified: u64,
+    path: OwnedObjectPath,
+}
+
+#[dbus_interface(name = "org.freedesktop.Secret.Collection")]
+impl Collection {
+    pub async fn delete(&self, #[zbus(object_server)] object_server: &ObjectServer) -> ObjectPath {
+        let _ = object_server.remove::<Collection, _>(&self.path).await;
+        ObjectPath::default().into() // returning '/' until we figure out Prompt
+    }
+
+    pub async fn search_items(&self, attributes: HashMap<&str, &str>) -> Vec<Item> {
+        let items = match KEYRING.get().unwrap().search_items(&attributes).await {
+            Ok(i) => i,
+            Err(_) => todo!(),
+        };
+
+        items // currently returns Vec<oo7::portal::Item>, this should be a path
+    }
+
+    pub async fn create_item(
+        &mut self,
+        properties: Properties,
+        secret: &str,
+        replace: bool,
+    ) -> (ObjectPath, ObjectPath) {
+        let label = properties.attributes().unwrap().get("label").unwrap();
+        let mut attributes: HashMap<&str, &str> = Default::default();
+
+        for (key, value) in properties.attributes().unwrap().into_iter() {
+            attributes.insert(key.as_str(), value.as_str());
+        }
+
+        let _collection = KEYRING
+            .get()
+            .unwrap()
+            .create_item(label, attributes.clone(), secret, replace)
+            .await;
+
+        // lookup just created item
+        let lookup_item = KEYRING
+            .get()
+            .unwrap()
+            .lookup_item(&attributes)
+            .await
+            .unwrap()
+            .unwrap();
+        // make prompt to get the secret and set it with set_secret()
+
+        let created_item_path = ObjectPath::default().into(); // how to get path from _collection
+        let prompt = ObjectPath::default().into(); // temp Prompt
+
+        let mut attributes_for_new: HashMap<&str, &str> = Default::default();
+        for (key, value) in lookup_item.attributes() {
+            attributes_for_new.insert(key.as_str(), value);
+        }
+
+        // TODO map portal::Item to crate::Item and push it to items attribute
+        let item = item::Item::new(
+            &attributes_for_new,
+            lookup_item.secret().to_vec(),
+            lookup_item.label(),
+            lookup_item.created().as_secs(),
+            lookup_item.modified().as_secs(),
+            self.path().await,
+        )
+        .await;
+        self.items.push(item);
+
+        (created_item_path, prompt)
+    }
+
+    #[dbus_interface(property, name = "Items")]
+    async fn items(&self) -> Vec<ObjectPath> {
+        let mut paths: Vec<ObjectPath> = Vec::new();
+
+        for path in &self.items {
+            paths.push(path.path().await.into());
```

Comment:
I don't think this should be async at all. then you could just write `self.items.iter().map(|item| item.path()).collect()`

Reply:
Done, thanks!